source/lib.ts

```diff
@@ -0,0 +1,312 @@
+			response.status(options.statusCode).send(options.message),
+		onLimitReached: (_request, _response, _optionsUsed) => {},
+		...passedOptions,
+	}
+
+	// Ensure that the store passed implements the either the `Store` or `LegacyStore`
+	// interface
+	if (
+		typeof options.store.increment !== 'function' ||
+		typeof options.store.resetKey !== 'function' ||
+		(options.skipFailedRequests &&
+			typeof options.store.decrement !== 'function')
+	) {
+		throw new Error(
+			'An invalid store was passed. Please ensure that the store is a class that implements the `Store` interface.',
+		)
+	} else {
+		// Promisify the store, if it is not already
+		options.store = promisifyStore(options.store)
+	}
+
+	// Return the 'clean' options
+	return options
+}
+
+/**
+ * Just pass on any errors for the developer to handle, usually as a HTTP 500
+ * Internal Server Error.
+ *
+ * @param fn {Express.RequestHandler} - The request handler for which to handle errors
+ *
+ * @returns {Express.RequestHandler} - The request handler wrapped with a `.catch` clause
+ *
+ * @private
+ */
+const handleAsyncErrors =
+	(fn: Express.RequestHandler): Express.RequestHandler =>
+	async (
+		request: Express.Request,
+		response: Express.Response,
+		next: Express.NextFunction,
+	) => {
+		try {
+			await Promise.resolve(fn(request, response, next)).catch(next)
+		} catch (error: unknown) {
+			next(error)
+		}
+	}
+
+/**
+ *
+ * Create an instance of IP rate-limiting middleware for Express.
+ *
+ * @param passedOptions {Options} - Options to configure the rate limiter
+ *
+ * @returns {RateLimitRequestHandler} - The middleware that rate-limits clients based on your configuration
+ *
+ * @public
+ */
+const rateLimit = (
+	passedOptions?: Partial<Options>,
+): RateLimitRequestHandler => {
+	// Parse the options and add the default values for unspecified options
+	const options = parseOptions(passedOptions ?? {})
+	// Call the `init` method on the store, if it exists
+	if (typeof options.store.init === 'function') options.store.init(options)
+
+	// Then return the actual middleware
+	const middleware = handleAsyncErrors(
+		async (
+			request: Express.Request,
+			response: Express.Response,
+			next: Express.NextFunction,
+		) => {
+			// First check if we should skip the request
+			const skip = await Promise.resolve(options.skip(request, response))
+			if (skip) {
+				next()
+				return
+			}
+
+			// Create an augmented request
+			const augmentedRequest = request as AugmentedRequest
+
+			// Get a unique key for the client
+			const key = await Promise.resolve(options.keyGenerator(request, response))
+			// Increment the client's hit counter by one
+			const { totalHits, resetTime } = await options.store.increment(key)
+
+			// Get the quota (max number of hits) for each client
+			const retrieveQuota =
+				typeof options.max === 'function'
+					? options.max(request, response)
+					: options.max
+
+			const maxHits = await Promise.resolve(retrieveQuota)
```

Comment:
Do we still need `Promise.resolve(..)` here, or could it just be `await retrieveQuota`?

Reply:
I just checked and it could be just `await retrieveQuota`... I was surprised that worked. I have used `Promise.resolve` in all places just in case there is some weird edge though.